cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -240,11 +240,11 @@ struct SubtractCheckedDate32 {
   }
 };
 
-template <bool is_32bit, int64_t multiple>
+template <bool left_is_32bit, int64_t multiple>
 struct AddTimeDuration {
   template <typename T, typename Arg0, typename Arg1>
-  static enable_if_t<is_32bit, T> Call(KernelContext*, Arg0 left, Arg1 right,
-                                       Status* st) {
+  static enable_if_t<(!std::is_same<Arg0, Arg1>::value && left_is_32bit), T> Call(
+      KernelContext*, Arg0 left, Arg1 right, Status* st) {
     T result = arrow::internal::SafeSignedAdd(left, static_cast<T>(right));
```

Comment:
Why not `arrow::internal::SafeSignedAdd(static_cast<T>(left), static_cast<T>(right))` and then you don't have to bother with different signatures?

Reply:
Would double cast not be wasteful? Will change.

Reply:
Done. Would you suggest to do this for `SubtractTimeDuration` and `SubtractTimeDurationChecked` as well?

Reply:
Everywhere it's applicable.

Reply:
Done.